### PR TITLE
Add a "by newbie, for newbie" tutorial about the digitech_xc0324 device handler

### PR DESCRIPTION
**Pull request by Geoff99**
Hi @zuckschwerdt 

For information and safekeeping only, no need to action this pull request (yet).

It appears to be more complicated than I anticipated to create a pull request to add the "by a newbie, for a newbie" tutorial into the rtl_433 wiki (as I said I would do in the discussion in pull request #849 )

As an interim measure, so that it remains visible and doesn't get damaged while I am tidying up the code in that pull request, I have moved the tutorial `.md` file to it's own branch, and I am submitting this pull request containing only the tutorial.

Apart from renaming the file, I have made no changes to the contents.  Once the actual code has been finalised, I will come back to this and make any necessary updates so that the tutorial and final version of the `digitech_xc0324.c` code are consistent.  

I will also do some research into how to (simply) submit a pull request against the wiki itself.  **If** I can solve that I will do so, and close this pull request; if I can't, I will have to leave the tutorial in the source directory until someone else can move it to the wiki.

Hope this approach is OK.

### CODE REVIEW

**Comment by Geoff99** [comment]
Hi @zuckschwerdt 

Looking at all the activity on the repo whizzing by you have certainly been very busy recently.

I have done some more work on the newbie tutorial.  With all the improvements and changes that have been made, there was a bit more alteration necessary and possible than I first anticipated.

Amongst other things I have managed to clone the wiki (sub) repository to my forked copy of rtl_433.  What I have up there is still **very much** a work in progress (especially towards the end where I have just parked spare text that doesn't really fit any more.  It is still **way** too long as well (but shortening text is not my forte :-( ).  If you like (but no obligation) you can have a quick look at it on
[WIP : newbie tutorial take 2](https://github.com/Geoff99/rtl_433/wiki/digitech_xc0324.README#deciphering-digitech-xc-0324-transmissions-and-creating-a-rtl_433-decoder).  I plan / hope (subject to some unavoidable interruptions) to get a bit more tidying up done over the next few days.

My reason for getting in touch now is that while redrafting the tutorial so it uses and explains all the really nice features added to rtl_433, I have realised that 

* instead of my current approach of leaving various trace and debug messages in the digitech_xc0324 decoder, so a reader who wants to could replicate the various stages of the reverse engineering process,
* I could instead create the necessary csv files in advance, using the `decoder_output_...` functions in incremental steps, exactly as explained  in the current draft of the tutorial, place them in the wiki, and link to them from the tutorial.

This would leave the production decoder much cleaner, but would mean hosting several csv files on the wiki.  Before I went too much further, I thought I should get your reaction to this approach.

PS Still keen on hex plus bit patterns, but might be *just* able to work around that by having multiple csv files (one with hex, the next with bits).  Would still need byte spacers between the bits though.

PPS Will do some more work on the tutorial, but once I've got it to something I'm OK with, will park it and leave it you to do whatever you feel is necessary (up to and including ignoring, if you think the effort to fix exceeds the likely value and audience).

PPPS Another track entirely, I have tried to use http://triq.net/iqs several time.  It opens up OK, but does nothing (that I can see) when I drag drop or open a .cu8 file.  I'm wondering whether there is anything that might be triggering my firewall (I haven't had time to dig around and look yet).

**Comment by zuckschwerdt** [comment]
I'd say leave the debug stuff in but comment it out. A reasonable syntax highlighter will then let people  quickly graps what's going on.

Hex+bits on it's way It's just a nasty debug function, so there is no format to follow :p
I thought about a switch (or maybe always in debug mode) to add binary to all hex output then. Don't know if that's reasonable.

Maybe we can put the tutorial directly into the wiki here.

Thanks about the info on iqs (IQ Spectrogram). That's odd. It won't upload anything so no firewall should trigger. It's a pure javascript renderer in the browser. A spectrogram picture should appear near instantly after dropping a file. Maybe you can inspect the Console for errors or warnings? (search for Console or Web Console or Developer Tools to get info.)
What browser/version on what OS shows the problem?

**Comment by zuckschwerdt** [comment]
Ah, how about a `-M bits` option to output a `bits=…` with every `codes=…` on `decoder_output_`? That would be a clean and universal solution.

**Comment by Geoff99** [comment]
Hi @merbanan

I have accepted the invitation to become a collaborator.  I must say I was surprised, flattered and honoured by the invitation.

I will be very careful *not* to push anything at all until I understand more about the protocols for making changes, and also I am certain that I'm not going to break anything!  Is there anything I can / should read to get myself up to speed on this.  It's the very first time I been in this role, so I am not versed at all in what might be even basic conventions and polite ways of working on a repository owned by someone else (truth be told I have only submitted probably about 3 pull requests ever before now!).

I don't know anywhere near as much as you and @zuckschwerdt , but I may perhaps be able to tidy up this tutorial documentation and offer that to the project.

Thank you once again for trusting me with this role.

**Comment by zuckschwerdt** [comment]
The wiki is a separate repo, you can edit that from a web browser or for bigger updates clone the repo and push commits. Feel free to do that with any change or addition you think of or have already prepared, like you noted.

On the hex/bits topic, I need and love to hear how rtl_433 is useful and what can be improved. And even a very tailored single-purpose wish is useful to hear. I hope there can always manifest a generalized concept to support that wish. I'll extend the meta-data option over the next few days to support bits-debugging, high resolution timestamps and the like.

File: src/devices/digitech_xc0324.README.md [hunk 1]
```diff
@@ -0,0 +1,1070 @@
...
+   *   each bit is effectively a parity bit for correspondingly
+      positioned bit in the real message.
+
+
+#### Batch processing sets of test data files to produce csv file output.
+
+My debugging / deciphering strategy (copied from several helpful blog posts)
+involves :
+
+* saving a set of test files (using `rtl_433 -a -t` ) then
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: -S all

File: src/devices/digitech_xc0324.README.md [hunk 2]
```diff
@@ -0,0 +1,1070 @@
...
+ 
+The protocol was deciphered using captured pulses from two sensors.
+See the test data in [rtl_433_tests/tests/XC-0324](https://github.com/merbanan/rtl_433_tests/tree/master/tests/XC-0324).
+
+My older sensor transmits a "clean" pulse (ie a captured pulse, examined using
+Audacity, has pretty stable I and Q values, ie little phase wandering):
+
+My newer transmitter seems less stable (ie within a single pulse, the I and
+Q components of the pulse signal appear to "rotate" through several cycles).
+The `rtl_433 -A -D` output correctly guesses the older transmitter modulation
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: -A -vv

File: src/devices/digitech_xc0324.README.md [hunk 3]
```diff
@@ -0,0 +1,1070 @@
...
+
+
+#### Batch processing sets of test data files to produce csv file output.
+
+My debugging / deciphering strategy (copied from several helpful blog posts)
+involves :
+
+* saving a set of test files (using `rtl_433 -a -t` ) then
+* running my 'under development' device handler on the set of test files,
+  (repeatedly till I got every thing figured out), using `-D` or `-DD`
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: -vv -vvv -vvvv

File: src/devices/digitech_xc0324.README.md [hunk 4]
```diff
@@ -0,0 +1,1070 @@
...
+To make use of these debug outputs I have a bash script (called
+`exam.sh`), located in the sub-directory where I have stored the test
+files.
+
+exam.sh
+
+```
+#! /bin/bash
+    
+for f in g*.cu8;
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: `rtl_433 […] g*.cu8` works now.

File: src/devices/digitech_xc0324.README.md [hunk 5]
```diff
@@ -0,0 +1,1070 @@
...
+
+Then I:
+
+* open `xc0324.csv` in a spreadsheet package (eg Excel),
+* manually edit in the correct observed temperatures for each test file,
+* sort into observed temperature order, and
+* start looking for patterns.
+
+#### Technical Aside : Filename labels in the "debug to csv" output.
+
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: Update this to -K FILE?

File: src/devices/digitech_xc0324.README.md [hunk 6]
```diff
@@ -0,0 +1,1070 @@
...
+This is explanation is pretty basic **provided you've done it before**,
+but may be helpful if, like I was, you are new to this.
+
+After setting up your own local clone of the rtl_433 repository, and
+compiling `rtl_433` from source
+(see [rtl_433 : Building.md](https://github.com/merbanan/rtl_433/blob/master/BUILDING.md)),
+the gameplan is:
+
+* Find the transmission frequency
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: …but offset `-f` by e.g. 20 kHz

File: src/devices/digitech_xc0324.README.md [hunk 7]
```diff
@@ -0,0 +1,1070 @@
...
+XC0324).  I also checked it using the "waterfall" display from `SDR#`
+(a Software Defined Radio program that I had installed from earlier
+experiments with my RTL-SDR dongle). The "waterfall" display also
+confirmed that the two sensors each transmitted new readings every 60
+seconds.
+
+###### Capturing test files.
+
+The mechanics of this are easy as well - `mkdir` and then `cd` to your
+chosen test data sub-directory and run `rtl_433 -a -t -f <frequency>`.
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: -A -S all

File: src/devices/digitech_xc0324.README.md [hunk 8]
```diff
@@ -0,0 +1,1070 @@
...
+  * FSK (Frequency Shift Keying) - think FM radio stations, and
+  * OOK (On Off Keying) - think AM radio stations.
+
+"Keying" means "something" changes to signify a bit - think morse code.
+With FSK a frequency changes, with OOK an amplitude changes (usually a
+"pulse" starts or stops or continues or ...).
+
+This makes more sense if you can "look" at the radio signal - there are
+several programs that do exactly that - I used `Audacity`, and,
+following the instructions in the (translated) French blog entry,
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: Or use the tests/sigrok-open.sh on the sample signal

File: src/devices/digitech_xc0324.README.md [hunk 9]
```diff
@@ -0,0 +1,1070 @@
...
+these parameters from the `rtl_433 -A` Pulse Analyser output.
+
+```
+Registering protocol [110] "XC0324"
+Registered 1 out of 110 device decoding protocols
+Test mode active. Reading samples from file: g001_433.917M_250k.cu8
+Input format: CU8 IQ (2ch uint8)
+Detected OOK package	@ @0.000000s
+Analyzing pulses...
+Total count:  150,  width: 43809		(175.2 ms)
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: All output here is now in µs. Perhaps update the screenshot and the text

File: src/devices/digitech_xc0324.README.md [hunk 10]
```diff
@@ -0,0 +1,1070 @@
...
+>* add this file to `src/CMakeLists.txt` and `Makefile.am`.
+>
+>The `src/devices/<protocolname>.c` file will contain two main things:
+>
+>* the decoding function itself (we will come back to this), and
+>* a r_device structure configuring the device and indicating, among other
+>  things, the modulation (and therefore the demodulator to use) and a
+>  1st level of encoding.
+
+It is also necessary to make a small change to `include/rtl_433.h` as
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: Not needed anymore

File: src/devices/digitech_xc0324.README.md [hunk 11]
```diff
@@ -0,0 +1,1070 @@
...
+
+The `src/devices/newtemplates.c` includes a r_device structure appropriate to its
+own example device.  For the XC0324, based on the information above,
+the device structure looks like:
+
+```
+r_device xc0324 = {
+    .name           = "XC0324",
+    .modulation     = OOK_PULSE_PPM_RAW,
+    .short_limit    = 190*4, //(130 + 250)/2  * 4
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: Don't use the *4, use µs only

File: src/devices/digitech_xc0324.README.md [hunk 12]
```diff
@@ -0,0 +1,1070 @@
...
+do for you.
+
+Much of the development involved enhancing the debug messages (eventually
+arriving at the "debug to csv" approach explained elsewhere), which
+made it simpler and quicker to explore the XC0324 protocol itself.
+
+**`include/rtl_433_devices.h`**
+
+It was straightforward to add the definition of the new device handler,
+I just inserted an extra line in the file at the obvious place
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: …at the end of the list

File: src/devices/digitech_xc0324.README.md [hunk 13]
```diff
@@ -0,0 +1,1070 @@
...
+```
+	...
+	DECL(wt1024) \
+	DECL(xc0324)
+	...
+```
+
+**`src/CMakeLists.txt` and `Makefile.am`**
+
+Amending these two files was equally simple :
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: sorted alphabetically.

File: src/devices/digitech_xc0324.README.md [hunk 14]
```diff
@@ -0,0 +1,1070 @@
...
+```
+and
+```
+    ...
+    devices/x10_sec.c \
+    devices/xc0324.c
+    ...
+```
+
+**`include/rtl_433.h`**
```
Comment (zuckschwerdt) [inline src/devices/digitech_xc0324.README.md]: Not needed anymore, remove

**Comment by Geoff99** [comment]
Hi @zuckschwerdt 

Many thanks indeed for flagging all these API changes for me.

I am embarrassed that I've been "off the air" on this contribution for significantly longer than I intended - the other commitment that I had been avoiding, (but became unavoidable :-( ) and that I thought would only take me a day or two hit a snag, and I got lost down a rabbit hole trying to resolve it.  I'm hopeful I will sort it out soon, and then I'll be able to get some clear time to work on this again.  

When I do I will put all these in, rebase my version of rtl_433 and work through the tutorial step by step myself to be sure I've got up to date arguments, example outputs etc.  And also adjust it to use the nice "bits" metadata options throughout the tutorial as well :-) 

Will alert you as soon as I've done that, so you can comment on any aspect of the (by then) revised tutorial.  Hopefully in a couple of days (fingers crossed)

**Comment by zuckschwerdt** [comment]
Great! Thanks. I hope to get some time the next weeks to perhaps refine the text. Feel free to update any part of the wiki.

**Comment by Geoff99** [comment]
Hi @zuckschwerdt 

I'm doing a bit of experimentation, with a  view to drafting a (short!) wiki entry about how to view a cu8 file.  I've covered `Audacity`, and I've got `sigrok-open.sh` to work (turned out I needed to install pulseview  - I'll cover that in my wiki entry).

Picking up something mentioned earlier in this pull request, I'm still having trouble with IQ spectrogram.

> Thanks about the info on iqs (IQ Spectrogram). That's odd. It won't upload anything so no firewall should trigger. It's a pure javascript renderer in the browser. A spectrogram picture should appear near instantly after dropping a file. Maybe you can inspect the Console for errors or warnings? (search for Console or Web Console or Developer Tools to get info.)
> What browser/version on what OS shows the problem?


It works when I use the Chrome browser, but not with IE11 or Edge (for this work I'm using my laptop, which has Windows 10 Professional as the OS).  Nothing happens when I try to import a cu8 file.  I can get the exact browser versions, if that's helpful.  I've captured screenshots of the console messages - but I can't figure out how to post anyhting like that (well not easily anyway).  So pasted in below (minus all the syntax highlighting etc) is the text part of the console message (from the Edge browser attempt).  Most of the messages come from just opening triq.net/iqs.  Only the final few lines appear when I try to import a cu8 file.

Not urgent, just thought it might be helpful for you if you wanted to find out what's happening.

PS Chrome grumbles that http is not secure, but it does work despite that :-)

```
 HTML1300: Navigation occurred.
iqs (1,1)

 Creating AppCache with manifest: 'http://triq.net/appcache.appcache'.

2 HTML1509: Unmatched end tag.
iqs (161,3)

 Creating 4 render workers.

 Manifest doesn't exist on the server: 'http://triq.net/appcache.appcache'.

 HTTP404: NOT FOUND - The server has not found anything matching the requested URI (Uniform Resource Identifier).
GET - http://triq.net/appcache.appcache

 g001_433.92M_250k.cu8
app.js (1,8510)

 SCRIPT5001: Number expected
app.js (1,13043)
```

**Comment by Geoff99** [comment]
Hi @zuckschwerdt 

Just had a thought, and tried IQS in Chrome (where it works) in developer mode.

Posted below is what pops up.  Just guessing (based on the deprecation warning) , but it could be due to browsers slowly getting more picky about sites that are **not** https

```
[Deprecation] Application Cache is restricted to secure contexts. Please consider migrating your application to HTTPS, and eventually shifting over to Service Workers. See https://goo.gl/rStTGz for more details.
app.js:1 Creating 4 render workers.
app.js:1 g001_433.92M_250k.cu8
app.js:1 block_norm=512, -22.632129520188467 dB, window weight=183.3213100000001
app.js:1 colors=256, color_norm=-8.533333333333333
app.js:1 samples=65536 / stride=163.37688442211055
app.js:1 width=399 height=512
worker.js:1 worker render: 154.90ms
worker.js:1 worker render: 184.20ms
worker.js:1 worker render: 169.90ms
worker.js:1 worker render: 189.20ms
app.js:1 total_time_scaled=262.144, time_markers_step=55, time_per_pixel=1.522064208984375
app.js:1 fft process: 1497.34716796875ms
app.js:1 fft render: 1694.797119140625ms
app.js:1 cB_hist_mean = 202.75200000000007
```

**Comment by zuckschwerdt** [comment]
About the `appcache` issue: does it help if you open the site as http://triq.net/iqs/ (note the trailing slash or http://triq.net/iqs/index.html if needed)? But Appcache is not really needed.
Interesting about the "Number expected" -- the File API says [there should be a number](https://developer.mozilla.org/en-US/docs/Web/API/File/lastModified)…

**Comment by Geoff99** [comment]
Hi @zuckschwerdt 

I've just pushed an entry to the wiki about how to view a saved cu8 file.
See https://github.com/merbanan/rtl_433/wiki/View-a-saved-cu8-file

I've headed it as "under development" so anyone who stumbles across it will understand it is still subject to review.

Comments, suggestions and corrections most welcome

**Comment by zuckschwerdt** [comment]
That is really great! It will help so many users and hopefully improve quality of samples and questions!

**Comment by zuckschwerdt** [comment]
@Geoff99 I tried adding a number of PolyFills for IE, but soon realized IE is lacking too many things. I'll try this again another time perhaps.

**Comment by Geoff99** [comment]
@zuckschwerdt 

> @Geoff99 I tried adding a number of PolyFills for IE, but soon realized IE is lacking too many things. I'll try this again another time perhaps.

No worries, IE is deprecated, even by Microsoft I understand.  I continue to use it only out of habit, and because of a password manager plugin I use (migrating to another one is on my todo list)

**Comment by Geoff99** [comment]
Hi @zuckschwerdt 

>That is really great! It will help so many users and hopefully improve quality of samples and questions!

Big :-), really glad you think it will be useful.  I start from a low base on many things I find fascinating, so I rely heavily on good documentation to get myself up to speed and started.

`rtl_433` is such an impressive project, it's a pleasure to be able to contribute.

**Comment by Geoff99** [comment]
Hi @zuckschwerdt 
I rather suspect the problem with IQ spectrogram was at my end on my system.
I just got prompted to update java on my machine.  When it finished I tried IQS using the edge browser, and it worked perfectly, first time.
Apologies for having wasted your time :-(

**Comment by zuckschwerdt** [comment]
No worries, I really need to clean up and polish the IQ Spectrogram someday.

**Pull request closed without merging**